transports/quic/src/provider.rs

```diff
@@ -0,0 +1,54 @@
+// Copyright 2022 Protocol Labs.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use futures::Future;
+use std::{
+    io,
+    net::SocketAddr,
+    task::{Context, Poll},
+};
+
+#[cfg(feature = "async-std")]
+pub mod async_std;
+#[cfg(feature = "tokio")]
+pub mod tokio;
+
+// Wrapped socket for non-blocking I/O operations.
+pub trait Provider: Unpin + Send + Sized + 'static {
+    // Wrap a socket.
+    // Note: The socket must be set to non-blocking.
+    fn from_socket(socket: std::net::UdpSocket) -> io::Result<Self>;
+
+    // Receive a single datagram message.
+    // Returns the message and the address the message came from.
+    fn poll_recv_from(&mut self, cx: &mut Context<'_>) -> Poll<io::Result<(Vec<u8>, SocketAddr)>>;
+
+    // Set sending the packet on the socket.
+    // Since only one packet may be sent at a time, this may only be called if a preceding call
+    // to [`Provider::poll_send_flush`] returned [`Poll::Ready`].
```

Comment:
The issue is that `AsyncWrite::poll_write` will only by able to write X bytes, so you always need to get the number of written bytes from function and advance the buffer accordingly.

Doable, but not sure if necessary here?

Reply:
Its because the `async_std::UdpSocket` doesn't provide a poll-like interface for sending a message, but instead only a `async fn send_to`. So what we do is we create the future in `start_send` (and thus take ownership of the bytes to send) and then drive that future in `poll_send_flush`. With a single `poll_send(&mut self, bytes: &[u8], ...)` this would not work. Does that make sense?